extension/entitystore/serviceprovider.go

```diff
@@ -0,0 +1,283 @@
+// telemetry received by CloudWatch Agent.
+func (s *serviceprovider) addEntryForLogGroup(logGroupName LogGroupName, serviceAttr ServiceAttribute) {
+	s.logGroups[logGroupName] = serviceAttr
+}
+
+type serviceAttributeProvider func() ServiceAttribute
+
+// mergeServiceAttributes takes in a list of functions that create ServiceAttributes, in descending priority order
+// (highest priority first), and proceeds down the list until we have obtained both a ServiceName and an
+// EnvironmentName.
+func mergeServiceAttributes(providers []serviceAttributeProvider) ServiceAttribute {
+	ret := ServiceAttribute{}
+
+	for _, provider := range providers {
+		serviceAttr := provider()
+
+		if ret.ServiceName == "" {
+			ret.ServiceName = serviceAttr.ServiceName
+			ret.ServiceNameSource = serviceAttr.ServiceNameSource
+		}
+		if ret.Environment == "" {
+			ret.Environment = serviceAttr.Environment
+		}
+
+		if ret.ServiceName != "" && ret.Environment != "" {
+			return ret
+		}
+	}
+
+	return ret
+}
+
+// logFileServiceAttribute function gets the relevant service attributes
+// service name is retrieved based on the following priority chain
+//  1. Incoming telemetry attributes
+//  2. CWA config
+//  3. instance tags - The tags attached to the EC2 instance. Only scrape for tag with the following key: service, application, app
+//  4. IAM Role - The IAM role name retrieved through IMDS(Instance Metadata Service)
+func (s *serviceprovider) logFileServiceAttribute(logFile LogFileGlob, logGroup LogGroupName) ServiceAttribute {
+	return mergeServiceAttributes([]serviceAttributeProvider{
+		func() ServiceAttribute { return s.serviceAttributeForLogGroup(logGroup) },
+		func() ServiceAttribute { return s.serviceAttributeForLogFile(logFile) },
+		s.serviceAttributeFromImdsTags,
+		s.serviceAttributeFromIamRole,
+		s.serviceAttributeFromAsg,
+		s.serviceAttributeFallback,
+	})
+}
+
+func (s *serviceprovider) getServiceNameAndSource() (string, string) {
+	sa := mergeServiceAttributes([]serviceAttributeProvider{
+		s.serviceAttributeFromImdsTags,
+		s.serviceAttributeFromIamRole,
+		s.serviceAttributeFallback,
+	})
+	return sa.ServiceName, sa.ServiceNameSource
+}
+
+func (s *serviceprovider) serviceAttributeForLogGroup(logGroup LogGroupName) ServiceAttribute {
+	if logGroup == "" {
+		return ServiceAttribute{}
+	}
+
+	return s.logGroups[logGroup]
+}
+
+func (s *serviceprovider) serviceAttributeForLogFile(logFile LogFileGlob) ServiceAttribute {
+	if logFile == "" {
+		return ServiceAttribute{}
+	}
+
+	return s.logFiles[logFile]
+}
+
+func (s *serviceprovider) serviceAttributeFromImdsTags() ServiceAttribute {
+	if s.GetIMDSServiceName() == "" {
+		return ServiceAttribute{}
+	}
+
+	return ServiceAttribute{
+		ServiceName:       s.GetIMDSServiceName(),
+		ServiceNameSource: ServiceNameSourceResourceTags,
+	}
+}
+
+func (s *serviceprovider) serviceAttributeFromIamRole() ServiceAttribute {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
```

Comment:
Unnecessary lock. `GetIAMRole()` already has a lock.

Reply:
fixed in new commit